pallets/system/src/header_builder.rs

```diff
@@ -263,3 +264,207 @@ impl Decode for AvailExtrinsic {
+						140, 206, 53, 32, 28, 16, 86, 117, 26, 110, 154, 16, 5, 21, 218, 249
+					]
+				]);
+			}
+		}
+	}
+
+	#[test]
+	fn test_single_merkle_proof() {
+		let mut leaves: Vec<[u8; 32]> = Vec::new();
+		let mut interested_leaf_position: Option<usize> = None;
+		let empty_vec: Vec<[u8; 32]> = vec![];
+
+		let avail_data: Vec<Vec<u8>> = vec![[
+			52, 53, 52, 102, 102, 56, 48, 99, 48, 56, 56, 97, 97, 55, 102, 97, 98, 57, 101, 49,
+		]
+		.to_vec()];
+
+		for (pos, xt) in avail_data.iter().enumerate() {
+			leaves.push(Sha256::hash(&xt));
+
+			if pos == 0usize {
+				interested_leaf_position = Some(leaves.len() - 1);
+			}
+		}
+		if leaves.len() > 0 {
+			if let Some(position) = interested_leaf_position {
+				let data_tree = MerkleTree::<Sha256>::from_leaves(&leaves);
+				let proof = data_tree.proof(&[position]);
+				let root_proof = proof.proof_hashes().to_vec();
+				// here the proof is shown empty because the root itself is the proof as there is only one appdata extrinsic
+				assert_eq!(root_proof, empty_vec);
+			}
+		}
+	}
+
+	#[test]
+	fn test_no_data_merkle_proof() {
+		let mut leaves: Vec<[u8; 32]> = Vec::new();
+		let mut interested_leaf_position: Option<usize> = None;
+		let avail_data: Vec<Vec<u8>> = vec![];
+		let mut proof_gen = false;
+
+		for (pos, xt) in avail_data.iter().enumerate() {
+			leaves.push(Sha256::hash(&xt));
+
+			if pos == 0usize {
+				interested_leaf_position = Some(leaves.len() - 1);
+			}
+		}
+		if leaves.len() > 0 {
+			if let Some(position) = interested_leaf_position {
+				let data_tree = MerkleTree::<Sha256>::from_leaves(&leaves);
+				let proof = data_tree.proof(&[position]);
+				let root_proof = proof.proof_hashes().to_vec();
+				assert_eq!(root_proof, vec![[0u8; 32]]);
+				proof_gen = true;
+			}
+		}
+		//no proof is generated because there is no appdata extrinsic
+		assert_eq!(proof_gen, false, "No proof generated");
+	}
+
+	///using rs-merkle proof verify function
+	#[test]
+	fn verify_merkle_proof() {
+		let avail_data: Vec<Vec<u8>> = vec![
+			[
+				48, 51, 51, 49, 102, 97, 55, 51, 101, 101, 101, 99, 99, 98, 101, 52, 101, 50, 50,
+				53,
+			]
+			.to_vec(),
+			[
+				54, 48, 100, 101, 100, 49, 102, 53, 97, 98, 54, 55, 50, 97, 55, 49, 50, 55, 98, 97,
+			]
+			.to_vec(),
+			[
+				50, 98, 49, 49, 102, 49, 100, 100, 51, 57, 53, 53, 54, 102, 98, 50, 97, 98, 52, 50,
+			]
+			.to_vec(),
+		];
+
+		let leaves = avail_data
+			.iter()
+			.map(|xt| Sha256::hash(&xt))
+			.collect::<Vec<[u8; 32]>>();
+
+		let merkle_tree = MerkleTree::<Sha256>::from_leaves(&leaves);
+		let indices_to_prove = vec![0, 1];
+		let leaves_to_prove = leaves.get(0..2).ok_or("can't get leaves to prove").unwrap();
+
+		let proof = merkle_tree.proof(&indices_to_prove);
+		let root = merkle_tree
+			.root()
+			.ok_or("couldn't get the merkle root")
+			.unwrap();
+
+		assert!(proof.verify(root, &indices_to_prove, leaves_to_prove, leaves.len()));
+	}
+}
```

Comment:
Can we add cases where we ask for proof when: 
- tree has no leaf
- tree has single leaf
- tree has power of 2 leaves